shared/utils/rootItemsGetter.ts

```diff
@@ -0,0 +1,86 @@
+import { ContentItem } from 'kentico-cloud-delivery';
+import { IWebhookContentItem } from '../external/models';
+
+interface IContext {
+    readonly visitedItems: string[],
+    readonly rootItemCodenames: string[],
+    readonly newItemsToVisit: ContentItem[],
+    readonly allItems: ContentItem[],
+}
+
+export const getRootCodenamesOfSingleItem = (item: IWebhookContentItem, allItems: ContentItem[]): string[] => {
+    if (item.type === 'zapi_specification') {
+        return [item.codename];
+    }
+
+    return getRootParents(item.codename, allItems);
+};
+
+const getRootParents = (codename: string, allItems: ContentItem[]): string[] => {
+    let itemsToVisit = getDirectParents(codename, allItems);
+    const visitedItems = [];
+    const rootItemCodenames = [];
+
+    while (itemsToVisit.length > 0) {
+        const newItemsToVisit = [];
+
+        itemsToVisit.forEach((item) =>
+            processItem(item, { visitedItems, rootItemCodenames, newItemsToVisit, allItems }));
+
+        itemsToVisit = newItemsToVisit;
+    }
+
+    return rootItemCodenames;
+};
+
+const processItem = (itemToProcess: ContentItem, context: IContext): void => {
+    const itemCodename = itemToProcess.system.codename;
+
+    if (context.visitedItems.includes(itemCodename)) {
+        return;
+    }
+    context.visitedItems.push(itemCodename);
+
+    if (itemToProcess.system.type === 'zapi_specification') {
+        context.rootItemCodenames.push(itemCodename);
+    } else {
+        const parents = getDirectParents(itemCodename, context.allItems);
+        parents.forEach((item) => context.newItemsToVisit.push(item));
+    }
+};
+
+const getDirectParents = (codename: string, allItems: ContentItem[]): ContentItem[] =>
+    allItems.filter((item) =>
+        checkAllItemsInRichTextFields(item, codename) ||
+        checkAllItemsInLinkedItemsFields(item, codename));
```

Comment:
This is a place for customization, I believe. If you passed a method along with `rootElementType` that we used in this filter, you should have generic algoritm, shouldn't you?

Reply:
The 2 `checkAllItems` methods are generic already, the `rootElementType` is enough to have a generic algorithm.